Support for OPTiMaDe to v1.0.0 and improved structured-data specifications

tachyontraveler opened the pull request:
- OPTiMaDe Related
    1. Upgraded the optimade specification to v1.0.0 
    2. A new landing page added at oqmd.org/optimade
    3. upgraded Lark dependency


- Structured Data related
   1. Rewrote the entire structured data specification for better indexing by search engines.

hegdevinayi commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -2,28 +2,32 @@
 class ParserError(Exception):
-    pass
-
+    def __init__(self,message):
```
Why override the constructor? The way you use it below (`raise ParserError("Unknown parser grammar version : " + str(version))`) should be enough to display any message.

tachyontraveler replied:
reverted it back to the default constructor derived from the parent class

hegdevinayi commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -2,28 +2,32 @@
 
+def get_grammar_data():
+    parser_grammer = {}
+    for name in glob(os.path.join(os.path.dirname(__file__), "./grammar", "*.g")):
```
Use `os.path.dirname(__file__), "grammar", "*.g"))` (no need of the `./`)

Suggestion (maybe this can be the subject of a separate, later PR): we should consider replacing all the `os.path.dirname` and `os.path.join` with `pkg_resources.resource_filename` and related conventions. The latter approach is much more robust across OSes and packages (e.g. in Python egg distributions vs source).

tachyontraveler replied:
I fixed the unnecessary `./`. But transferring to `pkg_resources` could be for the next PR. It's definitely a good idea to deal with any possible issues with egg files and cross-OS robustness, but it seems like we should be extensive tests while implementing such changes. I'll open an issue and work on it soon

hegdevinayi commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -2,28 +2,32 @@
+        with open(name) as f:
+            ver = tuple(
+                int(n)
+                for n in re.findall(r"\d+", str(os.path.basename(name).split(".g")[0]))
```
Use `os.path.splitext()` instead?

tachyontraveler replied:
Made the changes

hegdevinayi commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -41,19 +45,34 @@ def parse(self, filter_):
             self.filter = filter_
             return self.tree
         except Exception as e:
```
Is there a particular type of error we are trying to handle here? Or types? Would be better than a generic exception.

tachyontraveler replied:
I re-wrote most of the error handling in the parser and transformer. Now the error reporting is in accordance with OPTiMaDe specs. i.e; all errors happened because of bad request from the user are reported with 400 in the response. Other errors are reported as 500. 
@hegdevinayi

hegdevinayi commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -41,19 +45,34 @@ def parse(self, filter_):
         except Exception as e:
-            #    print(e)
-            return
+            raise ParserError(e)
```
If you tried the constructor overriding for this particular case, I think it is OK here to just re-raise the previous exception instead of raising the `ParserError`.

tachyontraveler replied:
Made the necessary changes

hegdevinayi commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -64,97 +83,223 @@ def __init__(self):
+            
+    def expression_clause(self,children):
+        try:
+            children = [item for item in children if not item==None]
```
(i.e., don't use `==` to check if some object is `None`)

tachyontraveler replied:
Made the necessary changes

hegdevinayi commented on qmpy/utils/oqmd_optimade/grammar/optimade_info.json:
```diff
@@ -0,0 +1 @@
+{"data": {"type": "info", "id": "/", "attributes": {"api_version": "1.0", "available_api_versions": [{"url": "http://oqmd.org/optimade/", "version": "1.0"}, {"url": "http://oqmd.org/optimade/v1/", "version": "1.0"}], "formats": ["json", "xml", "yaml"], "entry_types_by_format": {"json": ["structures"], "xml": ["structures"], "yaml": ["structures"]}, "available_endpoints": ["structures", "info", "l ...
```
Would recommend formatting all JSON data files with `indent = 2` to be (a) visually more parse-able, and (b) consistent with the rest of the data files in OQMD. (Unless there's a good reason not to, of course.)

tachyontraveler replied:
Made the necessary changes

hegdevinayi commented on qmpy/utils/oqmd_optimade/grammar/optimade_links.json:
```diff
@@ -0,0 +1 @@
+{"data": [{"type": "links", "id": "index", "attributes": {"name": "OQMD", "description": "Open Quantum Materials Database", "base_url": "http://oqmd.org/optimade/", "homepage": "http://oqmd.org", "link_type": "root"}}, {"type": "links", "id": "optimade", "attributes": {"name": "Materials Consortia", "description": "List of OPTIMADE providers maintained by the Materials Consortia organisation", "b ...
```
Would recommend formatting all JSON data files with indent = 2 to be (a) visually more parse-able, and (b) consistent with the rest of the data files in OQMD. (Unless there's a good reason not to, of course.)

tachyontraveler replied:
I changed all the new JSON files to have indent=2

hegdevinayi commented on qmpy/utils/oqmd_optimade/grammar/optimade_structures.info.json:
```diff
@@ -0,0 +1 @@
+{"data": {"description": "a structures entry", "properties": {"nelements": {"description": "Number of elements", "sortable": "false", "type": "integer"}, "lattice_vectors": {"description": "Unit cell lattice vectors", "unit": "Ao", "sortable": "false", "type": "list"}}, "formats": ["json", "xml", "yaml"], "output_fields_by_format": {"json": ["nelements", "lattice_vectors"], "xml": ["nelements"]}} ...
```
Would recommend formatting all JSON data files with indent = 2 to be (a) visually more parse-able, and (b) consistent with the rest of the data files in OQMD. (Unless there's a good reason not to, of course.)

tachyontraveler replied:
Made the necessary changes

hegdevinayi commented on qmpy/utils/oqmd_optimade/optimade_spec_data.py:
```diff
@@ -0,0 +1,22 @@
+
+def get_optimade_data(label):
+    if label=="info":
+        filename = os.path.join(os.path.dirname(__file__), "./grammar", "optimade_info.json")
```
Same comment as earlier re: the `os.path.join`.

tachyontraveler replied:
Made the necessary changes

hegdevinayi commented on qmpy/utils/oqmd_optimade/optimade_spec_data.py:
```diff
@@ -0,0 +1,22 @@
+    if label=="info":
+        filename = os.path.join(os.path.dirname(__file__), "./grammar", "optimade_info.json")
+        data  =json.load(open(filename,"r"))
+        return json.dumps(data,indent=4)
```
Would recommend formatting all JSON data files with indent = 2 to be (a) visually more parse-able, and (b) consistent with the rest of the data files in OQMD. (Unless there's a good reason not to, of course.)

tachyontraveler replied:
Made the necessary changes

hegdevinayi commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -64,97 +83,223 @@ def __init__(self):
+        return self.eq(a,b)
+    
+    def has_all(self,a,b):
+        assert type(b)==list
```
`assert isinstance(b, list)`

tachyontraveler replied:
Made the necessary changes

hegdevinayi commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -64,97 +83,223 @@ def __init__(self):
+        return _q
+    
+    def has_any(self,a,b):
+        assert type(b)==list
```
`assert isinstance(b, list)`

hegdevinayi replied:
Won't repeat this comment everywhere else but please do change all occurrences.

tachyontraveler replied:
Made the necessary changes

hegdevinayi reviewed (commented):
Great job upgrading to Optimade v1.0.0!

A couple of general notes:
- I think we should seriously consider adding some minimal tests for these (quite extensive!) set of functionalities implemented in the RESTful API. Obviously, the higher the code coverage the better but it would be really good to at least test most of the core functionality.
- Consider adding minimal docstrings to methods whose function is not immediately obvious from the name and/or the arguments. As a quick example, in the `QueryLarkDjangoParser.py` the `has_all`, `has_any`, `has_only` method names are not informative (by design), and that's ok but unless I actually read the logic it is hard to say what they might be implementing. So, a short docstring would greatly enhance readability.
- Please run all modified files through `black` for style consistency.

Side note/question: do you want to set up a `pre-commit` for qmpy with `black`. That will help automatically run all changes to python code through `black` before committing. It should take only a few minutes to set up.

tachyontraveler commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -64,97 +83,223 @@ def __init__(self):
+        return children[0]
+    
+    def property_first_comparison(self,children):
+        try:
```
removed all the unnecessary try-except conditions

tachyontraveler commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -64,97 +83,223 @@ def __init__(self):
+        
+    def expression(self,children):
+        try:
+            children = [item for item in children if not item==None]
```
Made the necessary changes

tachyontraveler commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -41,19 +45,34 @@ def parse(self, filter_):
+        return repr(self.lark)
 
+    
+class Property:
```
Made the necessary changes

tachyontraveler commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -64,97 +83,223 @@ def __init__(self):
+        return [self.opers[children[0]], children[1]]
+    
+    def set_op_rhs(self,children):
+        try:
```
Made the necessary changes

tachyontraveler commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -64,97 +83,223 @@ def __init__(self):
+    def value_list(self,children):
+        return list(children)
+        
+    def value(self,valTree):
```
Made the necessary changes

tachyontraveler commented on qmpy/utils/oqmd_optimade/QueryLarkDjangoParser.py:
```diff
@@ -64,97 +83,223 @@ def __init__(self):
+        assert len(valTree[0].children)==1
+        return valTree[0].children[0].value
+    
+    def _property(self,propTree):
```
Made the necessary changes

tachyontraveler commented:
Thanks a lot for your extensive review! I have made all the changes necessary, except the one with changing `os.path` to `pkg_resources`. That would need another dedicated PR, imp. It'd be great if you could go through my changes when you can - especially my replies on the unresolved conversations above. 


Also, I didn't notice the conflicting files before. It happened because I created this branch from `master`. If this PR looks good, can you accept the similar PR to `master` first, before accepting this PR? It's here: 
[https://github.com/wolverton-research-group/qmpy/pull/112](https://github.com/wolverton-research-group/qmpy/pull/112)

Once I merge the PR onto `master`, I can resolve the conflicts and re-request a review here for `production`. 
(Sorry about the mess that happened with these two PRs. I'm still not sure what's the best way to maintain `production` and `master` with the least amount of conflicts)

tachyontraveler commented:
> * I think we should seriously consider adding some minimal tests for these (quite extensive!) set of functionalities implemented in the RESTful API. Obviously, the higher the code coverage the better but it would be really good to at least test most of the core functionality.

I've added a set of tests to cover every part of the parser+transformer system, as far as I can think of. 

> * Consider adding minimal docstrings to methods whose function is not immediately obvious from the name and/or the arguments. As a quick example, in the `QueryLarkDjangoParser.py` the `has_all`, `has_any`, `has_only` method names are not informative (by design), and that's ok but unless I actually read the logic it is hard to say what they might be implementing. So, a short docstring would greatly enhance readability.

Added the docstrings on most of the classes and methods

> * Please run all modified files through `black` for style consistency.

Done!

> Side note/question: do you want to set up a `pre-commit` for qmpy with `black`. That will help automatically run all changes to python code through `black` before committing. It should take only a few minutes to set up.

I never knew that a `pre-commit` system existed lol. I'll into it during the next PR as I might have to read and test a bit on it to not mess it up lol

hegdevinayi commented:
Great job adding the docstrings and tests! @tachyontraveler 

I have just a few more comments/suggestions that I left on https://github.com/wolverton-research-group/qmpy/pull/112. Once those changes are in place/conversations resolved, I will approve that PR and leave this open.

Re: a separate PR for `pkg_resources`, that's completely fine. Also the `pre-commit` thing is very easy to setup, so let me know if you want chat on Slack about it.

tachyontraveler commented:
I've resolved the conflict. And yes, I'm down to setting up the pre-commit system ASAP, if it's that quick.

hegdevinayi reviewed (approved):
LGTM.
